### PR TITLE
Add TorchDataLoader to Train Benchmark

### PR DESCRIPTION
**Pull request by srinathk10**
## Why are these changes needed?
Add TorchDataLoader to Train Benchmark.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/train_tests/benchmark/train_benchmark.py [hunk 1]
```diff
@@ -178,48 +217,92 @@ def validate(self) -> Dict[str, float]:
...
+            return {"validation/loss": total_loss.item() / num_rows}
+
+        except Exception as e:
+            logger.error(f"[Validation] Error during validation: {str(e)}")
+            return {
+                "validation/loss": float("inf")
+            }  # Return infinity to indicate invalid validation
 
     def validate_step(self, input_batch, labels):
+        device = next(self.model.parameters()).device  # Ensure we get model's device
```
Comment (raulchen) [inline release/train_tests/benchmark/train_benchmark.py:252]: same as above.

**Review by raulchen — APPROVED** [review]

File: release/train_tests/benchmark/image_classification/factory.py [hunk 2]
```diff
@@ -88,11 +91,26 @@ def collate_fn(self, batch):
         return batch["image"], batch["label"]
 
 
+class ImageClassificationTorchDataLoaderFactory(TorchDataLoaderFactory):
+    """Factory for creating PyTorch DataLoaders for image classification tasks."""
+
+    def __init__(self, benchmark_config: BenchmarkConfig):
+        train_urls = IMAGENET_PARQUET_SPLIT_S3_DIRS["train"]
+        val_urls = IMAGENET_PARQUET_SPLIT_S3_DIRS["train"]
```
Comment (raulchen) [inline release/train_tests/benchmark/image_classification/factory.py:99]: Should we just use `["val"]` here, instead of using `["train"]` with a limit?

File: release/train_tests/benchmark/dataloader_factory.py [hunk 3]
```diff
@@ -36,8 +369,264 @@ def get_ray_datasets(self) -> Dict[str, Dataset]:
...
+                    # Move batch to device
+                    images, labels = batch
+                    logger.info(
+                        f"[DataLoader] Worker {worker_rank} processing batch {batch_idx} "
+                        f"(shape: {images.shape}, time since last: {time_since_last_batch:.2f}s)"
+                    )
+
+                    transfer_start = time.time()
+                    images = images.to(device, non_blocking=True)
+                    labels = labels.to(device, non_blocking=True)
```
Comment (raulchen) [inline release/train_tests/benchmark/dataloader_factory.py:497]: maybe also add a flag to control pin_memory and non_blocking.

File: release/train_tests/benchmark/image_classification/factory.py (hunk 2 above)
Comment (raulchen) [reply release/train_tests/benchmark/image_classification/factory.py:99]: maybe it's a question for @justinvyu , since the existing ray data benchmark does the same

File: release/release_tests.yaml [hunk 4]
```diff
@@ -2118,11 +2118,21 @@
         timeout: 2000
         script: RAY_TRAIN_V2_ENABLED=1 python train_benchmark.py --task=image_classification --dataloader_type=ray_data --num_workers=16
 
+    - __suffix__: full_training_torch_dataloader
+      run:
+        timeout: 2000
+        script: RAY_TRAIN_V2_ENABLED=1 python train_benchmark.py --task=image_classification --dataloader_type=torch --num_workers=16
```
[the remark this answers is not in the document]
Comment (srinathk10) [reply release/release_tests.yaml:2124]: These are train workers. Did you mean pytorch workers?

File: release/train_tests/benchmark/image_classification/factory.py (hunk 2 above)
Comment (justinvyu) [reply release/train_tests/benchmark/image_classification/factory.py:99]: Oh, the problem was that the `"val"` files do not include the label which is needed for validation. The labels need to be added as a column.

Comment (justinvyu) [reply release/train_tests/benchmark/image_classification/factory.py:99]: I do a limit on the train set instead: https://github.com/ray-project/ray/blob/master/release/train_tests/benchmark/image_classification/factory.py#L56

File: release/train_tests/benchmark/config.py [hunk 5]
```diff
@@ -35,15 +38,21 @@ class BenchmarkConfig(BaseModel):
     dataloader_config: DataLoaderConfig = Field(
         default_factory=lambda: DataLoaderConfig(),
     )
+    num_torch_workers: int = 0
+    torch_dataloader_timeout_seconds: int = 300
+    torch_pin_memory: bool = True
+    torch_non_blocking: bool = True
```
Comment (justinvyu) [inline release/train_tests/benchmark/config.py:44]: I have this nested `DataLoaderConfig` class that swaps depending on the `dataloader_type` that is set. For example, we could add a `DataloaderType.TORCH_S3_DATALOADER` option for the one you're adding, then put all the torch dataloader specific arguments into a `TorchDataloaderConfig`.

That way, only the relevant configs are available when a certain dataloder type is set (ex: it doesn't make sense to access torch dataloader configs when using ray data)

https://github.com/ray-project/ray/blob/a3d571920ed181d25c281fbaf5e83615f9383347/release/train_tests/benchmark/config.py#L7-L19

File: release/train_tests/benchmark/train_benchmark.py [hunk 6]
```diff
@@ -142,11 +151,18 @@ def get_next_batch(self, dataloader):
 
     def train_step(self, input_batch, labels):
         self.model.train()
+        self.optimizer.zero_grad()  # 1. Clear FIRST
```
Comment (justinvyu) [inline release/train_tests/benchmark/train_benchmark.py:154]: Did it matter to clear first vs at the end?

File: release/train_tests/benchmark/dataloader_factory.py [hunk 7]
```diff
@@ -36,8 +521,268 @@ def get_ray_datasets(self) -> Dict[str, Dataset]:
         return {}
 
 
+class TorchDataLoaderFactory(BaseDataLoaderFactory, S3Reader):
```
Comment (justinvyu) [inline release/train_tests/benchmark/dataloader_factory.py:524]: We can move this out to a `torch_dataloader_factory.py` and also the ray data stuff to a `ray_dataloader_factory.py`

File: release/train_tests/benchmark/dataloader_factory.py [hunk 8]
```diff
@@ -36,8 +521,268 @@ def get_ray_datasets(self) -> Dict[str, Dataset]:
...
+        logger.info(f"[DataLoader] Worker {worker_rank} creating train dataloader")
+
+        dataloader_config = self.get_dataloader_config()
+        device = self._get_device()
+
+        # Create dataset and dataloader
+        train_ds = S3ParquetImageIterableDataset(
+            file_urls=self._get_file_urls(self.train_url),
+            random_transforms=True,
+        )
```
Comment (justinvyu) [inline release/train_tests/benchmark/dataloader_factory.py:700]: We can make this `TorchDataloaderFactory` class agnostic to the actual dataset by making this `train_ds` initialization happen in an abstract method `get_iterable_datasets`, similar to this:

https://github.com/ray-project/ray/blob/a3d571920ed181d25c281fbaf5e83615f9383347/release/train_tests/benchmark/dataloader_factory.py#L52-L59

Then the dataloader factory just wraps the underlying dataset object with the right settings.

File: release/train_tests/benchmark/image_classification/factory.py [hunk 9]
```diff
@@ -88,11 +91,26 @@ def collate_fn(self, batch):
...
+        super().__init__(
+            benchmark_config,
+            train_urls,
+            val_urls,
+            limit_validation_rows=benchmark_config.limit_validation_rows,
+        )
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/factory.py:105]: If we do the comment above, then we also don't need to pass the url through the init, but rather override the `get_iterable_datasets` method here.

File: release/train_tests/benchmark/dataloader_factory.py [hunk 10]
```diff
@@ -1,13 +1,498 @@
...
+        1. Distributes files among workers
+        2. Processes rows with image transforms
+        3. Converts to tensors
+        4. Respects row limits per worker
+        """
+        try:
+            # Get worker info for file distribution
+            worker_info = torch.utils.data.get_worker_info()
+            worker_id = worker_info.id if worker_info else 0
+            num_workers = worker_info.num_workers if worker_info else 1
```
Comment (justinvyu) [inline release/train_tests/benchmark/dataloader_factory.py:390]: Just to clarify, these are dataloader workers, not training workers?

Do we still need to shard urls at the training worker level?

File: release/train_tests/benchmark/dataloader_factory.py [hunk 11]
```diff
@@ -1,13 +1,498 @@
...
+        limit_rows_per_worker: Optional[int] = None,
+    ):
+        """Initialize the dataset.
+
+        Args:
+            file_urls: List of S3 URLs to load
+            random_transforms: Whether to use random transforms for training
+            limit_rows_per_worker: Maximum number of rows to process per worker (None for all rows)
+        """
+        S3Reader.__init__(self)
```
Comment (justinvyu) [inline release/train_tests/benchmark/dataloader_factory.py:338]: i think `super().__init__()` should still work

File: release/train_tests/benchmark/dataloader_factory.py (hunk 10 above)
Comment (justinvyu) [reply release/train_tests/benchmark/dataloader_factory.py:390]: i.e. with `ray.train.get_context().get_world_rank()`

File: release/train_tests/benchmark/train_benchmark.py (hunk 6 above)
Comment (raulchen) [reply release/train_tests/benchmark/train_benchmark.py:154]: I don't think it matters. but we found that in the old benchmark, zero_grad was wrongly put in between forward and backward passes.

File: release/release_tests.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (srinathk10) [reply release/release_tests.yaml:2124]: Picked default a 8 Torch Data loaders and 16 Ray Train workers.

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py [hunk 12]
```diff
@@ -0,0 +1,564 @@
...
+            if worker_info is None:
+                files_to_read = self.file_urls
+            else:
+                per_worker = max(1, len(self.file_urls) // num_workers)
+                start_idx = worker_id * per_worker
+                end_idx = (
+                    start_idx + per_worker
+                    if worker_id < num_workers - 1
+                    else len(self.file_urls)
+                )
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:407]: last worker can get a few extra files if len(file_urls) doesn't divide evenly into num_workers.

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+                    file_rows=[1] * len(file_urls),  # Dummy row counts
+                    worker_rank=worker_rank,
+                    num_workers=num_workers,
+                )
+
+            # Otherwise, collect metadata for balanced distribution
+            logger.info(
+                "[DataLoader] Collecting file metadata for balanced distribution"
+            )
+            file_urls, file_rows = self._collect_file_info(bucket, prefix)
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:283]: Why do we not collect file info in the no limit_rows case? What if one worker gets less than limit_rows total rows?

(this might be the same issue Hao mentions in the slack thread)

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+        )
+    except RuntimeError:
+        logger.info("[DataLoader] Multiprocessing start method already set")
+
+# AWS configuration
+AWS_REGION = "us-west-2"
+
+
+@ray.remote
+def _fetch_parquet_metadata(bucket: str, key: str, file_url: str) -> Tuple[str, int]:
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:36]: maybe this can be a 0 CPU task since it's not doing any work (just doing an HTTP request)

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py [hunk 15]
```diff
@@ -0,0 +1,564 @@
...
+
+    # Step 3: Parse Parquet metadata
+    metadata = pq.read_metadata(io.BytesIO(response["Body"].read()))
+    total_rows = metadata.num_rows
+
+    logger.info(f"[DataLoader] File {file_url} has {total_rows} rows")
+    return file_url, total_rows
+
+
+class S3Reader:
```
[the remark this answers is not in the document]
Comment (srinathk10) [reply release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:74]: I was running into Boto Client picking issue around SSL. That's the reason, I went for this multiple inheritance model.

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py (hunk 13 above)
Comment (srinathk10) [reply release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:283]: With limit_rows set, we skip collecting rows. Instead we just sort the files and distribute files evenly between Train workers.

Without limit rows set, we will get rows per file and then try to distribute files to minimize row count difference between files.

Let me check if getting row count is cheap with Parquet S3 select. I believe it takes a few seconds. So I can default always get rows per file and evenly distribute files between Train workers.

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py (hunk 12 above)
Comment (srinathk10) [reply release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:407]: With limit, we are fine, since Torch workers will return once limit is reached.

Without limit, we sort the files based on row count. I can make sure to pick the files round robin between torch workers. Good catch!

File: release/train_tests/benchmark/config.py [hunk 16]
```diff
@@ -39,11 +49,14 @@ class BenchmarkConfig(BaseModel):
     # Training
     num_epochs: int = 1
     skip_train_step: bool = False
+    train_step_anomaly_detection: bool = False
+    limit_training_rows: int = 350000
```
[the remark this answers is not in the document]
Comment (srinathk10) [reply release/train_tests/benchmark/config.py]: Without limit, Torch runs into a hang in the train step backward propagation. So I added the limit here for train as well.

**Review by raulchen — APPROVED** [review]

File: release/train_tests/benchmark/image_classification/factory.py [hunk 17]
```diff
@@ -88,11 +100,198 @@ def collate_fn(self, batch):
...
+        train_file_urls = self._get_file_urls(self.train_url)
+        train_ds = S3ParquetImageIterableDataset(
+            file_urls=train_file_urls,
+            random_transforms=True,
+            limit_rows_per_worker=limit_training_rows_per_worker,
+        )
+
+        # Report training dataset configuration
+        ray.train.report(
+            {
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/factory.py:180]: oh hmm do we actually need these reports? train throws away these reports that are reported w/o a checkpoint

File: release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+            preprocess_fn = get_preprocess_map_fn(
+                decode_image=True, random_transforms=self.random_transforms
+            )
+
+            # Distribute files among workers
+            if worker_info is None:
+                files_to_read = self.file_urls
+            else:
+                # Round-robin distribution between PyTorch workers
+                files_to_read = self.file_urls[worker_id::num_workers]
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/torch_parquet_image_iterable_dataset.py:392]: nice

**Review by justinvyu — APPROVED** [review]

File: release/train_tests/benchmark/image_classification/factory.py (hunk 17 above)
Comment (justinvyu) [reply release/train_tests/benchmark/image_classification/factory.py:180]: might be better to save as a release test artifact (like the test metrics)

Comment (srinathk10) [reply release/train_tests/benchmark/image_classification/factory.py:180]: Ah I am just using this for sync. Don't need to save the report for now.
